storage/allocator.go

```diff
@@ -13,75 +13,291 @@
+	}
+	return stores[0], nil
+}
+
+// ShouldRebalance returns whether the specified store is overweight
+// according to the cluster mean and should rebalance a range.
+func (a *allocator) ShouldRebalance(s *proto.StoreDescriptor) bool {
+	a.Lock()
+	defer a.Unlock()
+	sl := a.findStores(*s.CombinedAttrs())
+
+	if sl.capacity.mean > maxAvailCapacityThreshold {
+		return float64(s.Capacity.RangeCount) > sl.count.mean
+	}
+	return s.Capacity.PercentAvail() < sl.capacity.mean
+}
+
+// selectRandom chooses count random store descriptors which match
+// the required attributes, do not include any of the existing
+// replicas, and return true when passed to the provided filter.
+// If the supplied filter is nil, it is ignored.
+func (a *allocator) selectRandom(count int, required proto.Attributes, existing []proto.Replica,
+	filter func(*proto.StoreDescriptor, *stat, *stat) bool) ([]*proto.StoreDescriptor, error) {
+	var descs []*proto.StoreDescriptor
+	sl := a.findStores(required)
+	used := getUsedNodes(existing)
+
+	// Randomly permute available stores matching the required attributes.
+	for _, idx := range a.randGen.Perm(len(sl.stores)) {
+		// Skip used nodes.
+		if _, ok := used[sl.stores[idx].Node.NodeID]; ok {
+			continue
+		}
+		// Filter descriptor.
+		if filter != nil && !filter(sl.stores[idx], &sl.count, &sl.capacity) {
+			continue
 		}
+		// Add this store; exit loop if we've satisfied count.
+		descs = append(descs, sl.stores[idx])
+		if len(descs) >= count {
+			break
+		}
+	}
+	if len(descs) == 0 {
+		return nil, util.Errorf("no available stores for %s", required)
 	}
+	return descs, nil
+}
 
-	var capacitySeen float64
-	targetCapacity := a.rand.Float64() * capacityTotal
+// findStores returns a store list matching the required attributes.
+// Results are cached for performance.
+//
+// If it cannot retrieve a StoreDescriptor from the Store's gossip, it
+// garbage collects the failed key.
+//
+// TODO(embark, spencer): consider using a reverse index map from
+// Attr->stores, for efficiency. Ensure that entries in this map still
+// have an opportunity to be garbage collected.
+func (a *allocator) findStores(required proto.Attributes) *storeList {
+	if a.storeLists == nil {
+		a.storeLists = map[string]*storeList{}
+	}
+	key := required.SortedString()
+	if sl, ok := a.storeLists[key]; ok {
+		return sl
+	}
+	sl := &storeList{}
+	a.storeLists[key] = sl
+
+	updateStoreList := func(key string) {
+		storeDesc, err := storeDescFromGossip(key, a.gossip)
+		if err != nil {
+			// We can no longer retrieve this key from the gossip store,
+			// perhaps it expired.
+			delete(a.capacityKeys, key)
+		} else if required.IsSubset(*storeDesc.CombinedAttrs()) {
+			sl.Update(storeDesc)
+		}
+	}
 
-	// Walk through candidates, stopping when
-	// we've passed the capacity target.
-	for _, c := range candidates {
-		capacitySeen += c.Capacity.PercentAvail()
-		if capacitySeen >= targetCapacity {
-			return c, nil
+	if a.deterministic {
```

Comment:
It sorts the keys from the capacityKeys set so that the random picking permutation is deterministic assuming a deterministic seed to the rng. The order which sets are traversed is based on something else.